src/algebra/permutation_matrix.jl

```diff
@@ -0,0 +1,242 @@
+export PermutationMatrix, PT_X_P, PT_X_P!, getind ,getindices
+
+import Base: *
+import LinearAlgebra: transpose , inv, mul!
+
+@doc raw"""
+`PermutationMatrix(ind::Array{T})` creates a permutation matrix with ones at coordinates `(k, ind[k]) for k = 1:length(ind)`.
+
+A permutation matrix represents a matrix containing only zeros and ones, which basically permutes the vector or matrix it is multiplied with.
+These matrices A are constrained by:
+```math
+    A_{ij} = \{0, 1\}\\
+    ∑_{i} A_{ij} = 1\\
+    ∑_{j} A_{ij} = 1
+```
+An example is the 3-dimensional permutation matrix
+```math
+A = \begin{bmatrix} 1 & 0 & 0 \\ 0 & 0 & 1 \\ 1 & 0 & 0 \end{bmatrix}
+```
+"""
+struct PermutationMatrix{ T <: Int } <: AbstractMatrix{T}
+    ind::Vector{T}
+    function PermutationMatrix(ind::Vector{T}) where { T <: Int }
+        return new{T}(ind)
+    end
+end
+
+@doc raw"""
+`PermutationMatrix(dim::T; switch_first::Bool=true)` Generates a random permutation matrix of size (dim x dim).
+The `switch_first` argument specifies whether the first index always has to be permuted.
+"""
+function PermutationMatrix(dim::T; switch_first::Bool=true) where { T <: Int }
+    ind = shuffle(collect(1:dim))
+    if switch_first && ind[1] == 1
+        tmp = ind[2]
+        ind[2] = ind[1]
+        ind[1] = tmp
+    end
+    return PermutationMatrix(ind)
+end
+
+
+# extensions of base functionality
+Base.eltype(::PermutationMatrix{T}) where { T } = T
+Base.size(mat::PermutationMatrix)               = (length(mat), length(mat)) 
+Base.length(mat::PermutationMatrix)             = length(mat.ind)
```

Comment:
`Base.length(matrix::AbstractMatrix)` usually returns `prod(size(matrix))`. So for `3x3` matrix `length` should be equal to `9`.

Reply:
But be careful as I see you use `length` later on with a different assumption. Maybe its fine to keep it as it is now.

Reply:
I'd be very concerned about leaving as is. `length` is used in lots of places in Base, and having a nonstandard definition will very likely lead to weird behavior and subtle bugs down the road. If other code comes to depend on the nonstandard implementation, it can be very hard to back out later.

Reply:
Resolved by 0e88df9.